### PR TITLE
[FIX] fix issues related to image with a single or no map / label

### LINKED ISSUE #5907: [BUG] nilearn.plotting.plot_prob_atlas does not support atlas plotting with one single ROI
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Operating system

- [x] Linux
- [ ] Mac
- [ ] Windows

### Operating system version

Linux Ubuntu 22.04

### Python version

- [ ] 3.14
- [ ] 3.13
- [x] 3.12
- [ ] 3.11
- [ ] 3.10

### nilearn version

Dev version

### Expected behavior

The plot should work even if there is only one ROI

### Current behavior & error messages

This is what I got:

```bash
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/tmp/script.py in <module>
     21 
     22 from nilearn.plotting import plot_prob_atlas, show
---> 23 plot_prob_atlas(
     24     pos_regions_img,
     25     display_mode="z",

~/mygit/nilearn/nilearn/plotting/image/img_plotting.py in plot_prob_atlas(maps_img, bg_img, view_type, threshold, linewidths, cut_coords, output_file, display_mode, figure, axes, title, annotate, draw_cross, black_bg, dim, colorbar, cmap, vmin, vmax, alpha, radiological, **kwargs)
   1157             "segmented colors", color_list, n_maps + 1
   1158         )
-> 1159         display._show_colorbar(cmap, Normalize(1, n_maps + 1))
   1160         tick_locator = MaxNLocator(nbins=10)
   1161         display.locator = tick_locator

~/mygit/nilearn/nilearn/plotting/displays/_slicers.py in _show_colorbar(self, cmap, norm, cbar_vmin, cbar_vmax, threshold)
    729         self._colorbar_ax.set_facecolor("w")
    730 
--> 731         self._cbar = create_colorbar_for_fig(
    732             figure,
    733             self._colorbar_ax,

~/mygit/nilearn/nilearn/plotting/_engine_utils.py in create_colorbar_for_fig(fig, axes, cmap, norm, threshold, cbar_vmin, cbar_vmax, n_ticks, tick_format, spacing, orientation, threshold_color)
    174         tick_format=tick_format,
    175     )
--> 176     thrs_cmap = threshold_cmap(cmap, norm, threshold, threshold_color)
    177     bounds = get_cbar_bounds(cbar_vmin, cbar_vmax, thrs_cmap.N, tick_format)
    178 

~/mygit/nilearn/nilearn/plotting/_engine_utils.py in threshold_cmap(cmap, norm, threshold, threshold_color)
     52     """
     53     cmap = plt.get_cmap(cmap)
---> 54     cmaplist = [cmap(i) for i in range(cmap.N)]
     55 
     56     if threshold is not None:

~/mygit/nilearn/nilearn/plotting/_engine_utils.py in <listcomp>(.0)
     52     """
     53     cmap = plt.get_cmap(cmap)
---> 54     cmaplist = [cmap(i) for i in range(cmap.N)]
     55 
     56     if threshold is not None:

~/.local/lib/python3.10/site-packages/matplotlib/colors.py in __call__(self, X, alpha, bytes)
    738         RGBA values with a shape of ``X.shape + (4, )``.
    739         """
--> 740         rgba, mask = self._get_rgba_and_mask(X, alpha=alpha, bytes=bytes)
    741         if not np.iterable(X):
    742             rgba = tuple(rgba)

~/.local/lib/python3.10/site-packages/matplotlib/colors.py in _get_rgba_and_mask(self, X, alpha, bytes)
    769         """
    770         if not self._isinit:
--> 771             self._init()
    772 
    773         xa = np.array(X, copy=True)

~/.local/lib/python3.10/site-packages/matplotlib/colors.py in _init(self)
   1062     def _init(self):
   1063         self._lut = np.ones((self.N + 3, 4), float)
-> 1064         self._lut[:-3, 0] = _create_lookup_table(
   1065             self.N, self._segmentdata['red'], self._gamma)
   1066         self._lut[:-3, 1] = _create_lookup_table(

~/.local/lib/python3.10/site-packages/matplotlib/colors.py in _create_lookup_table(N, data, gamma)
    655 
    656     if x[0] != 0. or x[-1] != 1.0:
--> 657         raise ValueError(
    658             "data mapping points must start with x=0 and end with x=1")
    659     if (np.diff(x) < 0).any():

ValueError: data mapping points must start with x=0 and end with x=1
```

### Steps and code to reproduce bug

```python
rom nilearn.regions import connected_regions
from nilearn.image import threshold_img
from nilearn.plotting  import view_img

from nilearn.datasets import load_sample_motor_activation_image
map_img = load_sample_motor_activation_image()

percentile = 99.4
min_region_size = 3000

pos_thresholded_img = threshold_img(
        map_img,
        threshold=f"{percentile}%",
        copy=True,
        two_sided=False,
        copy_header=True,
    )
pos_regions_img, pos_index = connected_regions(
    pos_thresholded_img, min_region_size=min_region_size
)

from nilearn.plotting import plot_prob_atlas, show
plot_prob_atlas(
    pos_regions_img,
    display_mode="z",
    cut_coords=5,
)


```

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Closes #5907

Changes proposed in this pull request:

- [x] handle look up table generation when label image is empty
- [x] add systematic checks of maps maskers with one or more maps
- [x] add systematic checks of labels maskers with one or more labels 
- [x] add systematic checks of sphere maskers with one or more sphere   
- [x] raise error when fitting maskers initialised with empty map_image
  - [x] maps
    - [x] surface
    - [x] nifti
  - [x] labels
    - [x] surface
    - [x] nifti 
- [x] raise error when fitting maps maskers initialised with a mask_img that completely masks the map_image
  - [x] maps
    - [x] surface
    - [x] nifti
  - [x] labels
    - [x] surface
    - [x] nifti 
- [x] plot_prob_atlas with a single map image maps: do not use colorbar if the atlas has a single map
- [x] fix nifti sphere masker to properly handle set_output

### CODE REVIEW

File: nilearn/plotting/image/img_plotting.py [hunk 1]
```diff
@@ -1151,34 +1151,46 @@ def plot_prob_atlas(
...
-        )
+        if n_maps == 1:
+            warnings.warn(
+                (
+                    "\nThe image maps contains a single image."
+                    "\nNo color map needed."
+                ),
+                RuntimeWarning,
+                stacklevel=find_stack_level(),
+            )
```
Comment (Remi-Gau) [inline nilearn/plotting/image/img_plotting.py:1242]: maybe that's a drastic way to handle the problem but I believe the colorbar is not needed in this case

**Review by bthirion — COMMENTED** [review]
LGTM so far !

**Comment by Remi-Gau** [comment]
OK also encountering the issue of maps / label images with no maps / images in them...

**Comment by bthirion** [comment]
Do you want to solve al cases in this PR ?

**Comment by Remi-Gau** [comment]
> Do you want to solve al cases in this PR ?

yup: added a list of all the cases I could think of in the top comment of the PR

**Comment by bthirion** [comment]
LMK when you want a review on it.

**Comment by bthirion** [comment]
The `test_masker_attributes_with_fit_masker` seems to be failing.

**Comment by Remi-Gau** [comment]
> The `test_masker_attributes_with_fit_masker` seems to be failing.

yes the tests in CI have so much red and green that it looks like a xmas tree

File: nilearn/maskers/nifti_spheres_masker.py [hunk 2]
```diff
@@ -681,13 +683,18 @@ def transform_single_imgs(self, imgs, confounds=None, sample_mask=None):
...
             memory_level=self.memory_level,
             # kwargs
             verbose=self.verbose,
         )
-        return np.atleast_1d(signals)
+
+        if self.n_elements_ == 1:
+            return signals
+        else:
+            return np.atleast_1d(signals)
```
Comment (Remi-Gau) [inline nilearn/maskers/nifti_spheres_masker.py:644]: testing with 1 or 3 spheres showed that we had some issues on when using set_output for this masker

**Review by bthirion — APPROVED** [review]
LGTM, thx !

File: nilearn/decomposition/tests/test_decomposition_estimators.py [hunk 3]
```diff
@@ -336,50 +359,10 @@ def test_pass_masker_arg_to_estimator(
     check_decomposition_estimator(est, data_type)
 
 
-# TODO passing confounds does not affect output with CanICA, DictLearning
-# @pytest.mark.parametrize("estimator", [CanICA, _MultiPCA, DictLearning])
-@pytest.mark.slow
-@pytest.mark.parametrize("estimator", [_MultiPCA])
-@pytest.mark.parametrize("data_type", ["nifti", "surface"])
-def test_with_confounds(
```
Comment (Remi-Gau) [inline nilearn/decomposition/tests/test_decomposition_estimators.py:343]: moved up

File: nilearn/decomposition/tests/test_decomposition_estimators.py [hunk 4]
```diff
@@ -173,86 +173,94 @@ def test_masker_attributes_with_fit_masker(
...
-        AssertionError, assert_array_equal, signals, signals_confounds
+        AssertionError, assert_array_equal, components, components_clean
     )
 
 
 @pytest.mark.slow
 @pytest.mark.parametrize("estimator", [CanICA, DictLearning])
 @pytest.mark.parametrize("data_type", ["nifti", "surface"])
-def test_transform_single_image(
+@pytest.mark.parametrize("single_subject", [True, False])
```
Comment (Remi-Gau) [inline nilearn/decomposition/tests/test_decomposition_estimators.py:251]: merged 2 test by using parametrization

**Comment by Remi-Gau** [comment]
Note that this PR also fixes some of our tests that would lead to using emtpy label / maps image.

**Comment by Remi-Gau** [comment]
Several tests for DictLearnng are now failing

https://github.com/nilearn/nilearn/actions/runs/20438255438/job/58724853904?pr=5908#step:8:21519

This is because the componenent image found by DictLearning is actually an empty image map.


```
data_type = 'surface', canica_data = [<SurfaceImage (115, 5)>, <SurfaceImage (115, 5)>, <SurfaceImage (115, 5)>], estimator = <class 'nilearn.decomposition.dict_learning.DictLearning'>, single_subject = True

    @pytest.mark.slow
    @pytest.mark.parametrize("estimator", [CanICA, DictLearning])
    @pytest.mark.parametrize("data_type", ["nifti", "surface"])
    @pytest.mark.parametrize("single_subject", [True, False])
    def test_transform(
        data_type,  # noqa: ARG001
        canica_data,
        estimator,
        single_subject,
    ):
        """Test transform and inverse transform.
    
        Passing a single image instead of list to transform should work
        but still return a list.
        """
        est = estimator(
            n_components=3,
            random_state=RANDOM_STATE,
            smoothing_fwhm=None,
            standardize="zscore_sample",
        )
    
        if single_subject:
            data = canica_data[0]
            assert not isinstance(data, list)
        else:
            data = canica_data
    
>       est.fit(data)

canica_data = [<SurfaceImage (115, 5)>, <SurfaceImage (115, 5)>, <SurfaceImage (115, 5)>]
data       = <SurfaceImage (115, 5)>
data_type  = 'surface'
est        = DictLearning(n_components=3, random_state=42, smoothing_fwhm=None,
             standardize='zscore_sample')
estimator  = <class 'nilearn.decomposition.dict_learning.DictLearning'>
single_subject = True

nilearn/decomposition/tests/test_decomposition_estimators.py:281: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
nilearn/decomposition/_base.py:578: in fit
    self.maps_masker_.fit()
        confounds  = None
        data       = array([[-9.55699168e-01, -6.43532653e-01, -1.82659210e+00,
        -1.58184634e+00,  7.42533537e-01,  8.73602145e-02,
...-1.51810367e+00,  5.58348606e-01,
        -1.27979665e+00, -1.14883506e+00, -7.54101208e-01,
         7.42445764e-02]])
        imgs       = [<SurfaceImage (115, 5)>]
        maps_masker_kwargs = {'memory': None, 'memory_level': 0}
        masker_type = 'multi_surface'
        self       = DictLearning(n_components=3, random_state=42, smoothing_fwhm=None,
             standardize='zscore_sample')
nilearn/maskers/surface_maps_masker.py:219: in fit
    return self._fit(imgs)
        imgs       = None
        self       = SurfaceMapsMasker(maps_img=<SurfaceImage (115, 3)>,
                  mask_img=<SurfaceImage (115,)>, memory_level=0)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = SurfaceMapsMasker(maps_img=<SurfaceImage (115, 3)>,
                  mask_img=<SurfaceImage (115,)>, memory_level=0), imgs = None

    def _fit(self, imgs):
        if self.maps_img is None:
            raise ValueError(
                "Please provide a maps_img during initialization. "
                "For example, masker = SurfaceMapsMasker(maps_img=maps_img)"
            )
    
        if imgs is not None:
            check_surf_img(imgs)
    
        self._fit_cache()
    
        mask_logger("load_regions", self.maps_img, verbose=self.verbose)
    
        # check maps_img data is 2D
        self.maps_img.data._check_ndims(2, "maps_img")
    
        maps_data = get_data(self.maps_img)
        if np.all(maps_data == 0):
>           raise ValueError("maps_img contains no map.")
E           ValueError: maps_img contains no map.

imgs       = None
maps_data  = array([[-0.,  0., -0.],
       [ 0., -0.,  0.],
       [-0., -0., -0.],
       [ 0., -0., -0.],
       [-0.,  0.,  0.]...      [-0., -0., -0.],
       [-0.,  0., -0.],
       [ 0.,  0.,  0.],
       [-0., -0., -0.],
       [ 0.,  0.,  0.]])
self       = SurfaceMapsMasker(maps_img=<SurfaceImage (115, 3)>,
                  mask_img=<SurfaceImage (115,)>, memory_level=0)

nilearn/maskers/surface_maps_masker.py:240: ValueError
```

Do we agree that in this case an error should be raised?

If so maybe we should have a different message in the error because the current message (' maps_img contains no map.') may not help the user much.

This also means that for a long time we have been testing DictLearning with some 'empty data'.

**Comment by Remi-Gau** [comment]
Removing this PR from the 0.13 milestone: still too many loose ends

**Comment by bthirion** [comment]
Indeed, my intention was not to make this a  high priority thing.
Thx for the effort !

**Review by bthirion — APPROVED** [review]
LGTM.

**Comment by Remi-Gau** [comment]
OK for this was we are going to generate data that actually contains meaningful data that can be fitted with DictLearning because all the dummy data we are throwing at it ATM fail

**Comment by bthirion** [comment]
> OK for this was we are going to generate data that actually contains meaningful data that can be fitted with DictLearning because all the dummy data we are throwing at it ATM fail

Makes sense actually ;-)

**Comment by Remi-Gau** [comment]
note that https://github.com/nilearn/nilearn/pull/5908/commits/f6d9dad353b161a8f263a2ee3646fc7a62c6fb4c was generated by claude

File: nilearn/_utils/tests/test_bids.py [hunk 5]
```diff
@@ -65,3 +65,21 @@ def test_check_look_up_table_errors(shape_3d_default):
         check_look_up_table(
             lut=lut, atlas=mock_regions_with_missing_labels, strict=True
         )
+
+
+def test_check_look_up_table_empty_image_error(img_3d_zeros_eye):
```
Comment (bthirion) [inline nilearn/_utils/tests/test_bids.py:74]: Do you mind adding a one-liner here (although the title is already quite explicit) ?

File: nilearn/decomposition/tests/conftest.py [hunk 6]
```diff
@@ -298,32 +302,65 @@ def _canica_components_volume(shape):
...
+
+    n_timepoints : int
+        Number of timepoints. 200 has been empirically validated to produce
+        SVD-reduced features that exceed DictLearning's default alpha=10.
+    """
+    n_components = components.shape[0]
+    n_left = mesh.parts["left"].coordinates.shape[0]
+
+    weights = rng.normal(size=(n_timepoints, n_components))
+    data_all = weights @ components + 0.01 * rng.normal(
```
Comment (bthirion) [inline nilearn/decomposition/tests/conftest.py:336]: Here we have a hard coded hyperparameter ; maybe give control on it to users ?

**Review by bthirion — COMMENTED** [review]
LGTM overall, thx !

File: nilearn/_utils/tests/test_bids.py (hunk 5 above)
Comment (Remi-Gau) [reply nilearn/_utils/tests/test_bids.py:74]: this is strange: the lack of docstring should have been flagged by precommit/ruff

File: nilearn/decomposition/tests/conftest.py (hunk 6 above)
Comment (Remi-Gau) [reply nilearn/decomposition/tests/conftest.py:336]: I will add a weights=None default, in case we want to pass something more explicit

File: nilearn/_utils/tests/test_bids.py (hunk 5 above)
Comment (Remi-Gau) [reply nilearn/_utils/tests/test_bids.py:74]: ah yeah it's in _utils so it is NOT a public function hence why it is not caught

Comment (Remi-Gau) [reply nilearn/_utils/tests/test_bids.py:74]: done

File: nilearn/decomposition/tests/conftest.py (hunk 6 above)
Comment (Remi-Gau) [reply nilearn/decomposition/tests/conftest.py:336]: done

**Comment by Remi-Gau** [comment]
@hndgzkn 
we are getting a 403 error on one of the 'new' freesurfer tests

nilearn/datasets/tests/test_struct.py::test_fetch_surf_fsaverage_wrong_order[fsaverage4] - requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://files.de-1.osf.io/v1/resources/mk53n/providers/osfstorage/641dc92da69c1b01698c0b32

this is because the test is actually trying to download some data and tests are not supposed to try to do that to avoid this kind of problem

we may have to find some work-around for the new tests we have added regarding vertices order

**Comment by Remi-Gau** [comment]
> @hndgzkn we are getting a 403 error on one of the 'new' freesurfer tests
> 
> nilearn/datasets/tests/test_struct.py::test_fetch_surf_fsaverage_wrong_order[fsaverage4] - requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://files.de-1.osf.io/v1/resources/mk53n/providers/osfstorage/641dc92da69c1b01698c0b32
> 
> this is because the test is actually trying to download some data and tests are not supposed to try to do that to avoid this kind of problem
> 
> we may have to find some work-around for the new tests we have added regarding vertices order

OK for now I will mark the test as flaky

**Comment by hndgzkn** [comment]
> we may have to find some work-around for the new tests we have added regarding vertices order

This test will actually be removed after updating datasets on OSF, so may be we do not need to spend time to find a workaround.

Or, for such tests, could we imagine a solution as creating a mark not to run on CI?

**Comment by Remi-Gau** [comment]
> This test will actually be removed after updating datasets on OSF, so may be we do not need to spend time to find a workaround.
> 
> Or, for such tests, could we imagine a solution as creating a mark not to run on CI?

yup but I think some of those tests will remain: like the ones that do a comparison of the plotted freesurfer data for different mesh resolution
